File: src/main/java/org/cubexmc/metro/physics/KinematicLeadStateMath.java
package org.cubexmc.metro.physics;

import org.bukkit.Location;
import org.bukkit.util.Vector;
import org.cubexmc.metro.util.LocationUtil;

final class KinematicLeadStateMath {

    static final class LeadState {
        final double x;
        final double y;
        final double z;
        final double vx;
        final double vy;
        final double vz;
        final Location location;

        LeadState(double x, double y, double z, double vx, double vy, double vz, Location location) {
            this.x = x;
            this.y = y;
            this.z = z;
            this.vx = vx;
            this.vy = vy;
            this.vz = vz;
            this.location = location;
        }
    }

    private KinematicLeadStateMath() {
    }

    static LeadState advanceAndRecover(Location currentLocation, Vector correctedVelocity, double timeFraction) {
        Vector step = correctedVelocity.clone().multiply(timeFraction);

        double nextX = currentLocation.getX() + step.getX();
        double nextY = currentLocation.getY() + step.getY();
        double nextZ = currentLocation.getZ() + step.getZ();

        Location nextLocation = new Location(currentLocation.getWorld(), nextX, nextY, nextZ);
        if (!LocationUtil.isRail(nextLocation)) {
            Location snapped = LocationUtil.snapToRail(nextLocation, currentLocation.getWorld());
            if (snapped != null && snapped.distanceSquared(nextLocation) < 4.0) {
                nextLocation = snapped;
                nextX = snapped.getX();
                nextY = snapped.getY();
                nextZ = snapped.getZ();
            }
        }

        nextLocation = RailPathUtil.project(nextLocation);
        nextX = nextLocation.getX();
        nextY = nextLocation.getY();
        nextZ = nextLocation.getZ();

        return new LeadState(
                nextX,
                nextY,
                nextZ,
                correctedVelocity.getX(),
                correctedVelocity.getY(),
                correctedVelocity.getZ(),
                nextLocation);
    }
}

File: src/main/java/org/cubexmc/metro/physics/KinematicLeadStateMath.kt
package org.cubexmc.metro.physics

import org.bukkit.Location
import org.bukkit.util.Vector
import org.cubexmc.metro.util.LocationUtil

internal object KinematicLeadStateMath {

    internal class LeadState(
        @JvmField val x: Double,
        @JvmField val y: Double,
        @JvmField val z: Double,
        @JvmField val vx: Double,
        @JvmField val vy: Double,
        @JvmField val vz: Double,
        @JvmField val location: Location,
    )

    @JvmStatic
    fun advanceAndRecover(currentLocation: Location, correctedVelocity: Vector, timeFraction: Double): LeadState {
        val step = correctedVelocity.clone().multiply(timeFraction)

        var nextX = currentLocation.x + step.x
        var nextY = currentLocation.y + step.y
        var nextZ = currentLocation.z + step.z

        var nextLocation = Location(currentLocation.world, nextX, nextY, nextZ)
        if (!LocationUtil.isRail(nextLocation)) {
            val snapped = LocationUtil.snapToRail(nextLocation, currentLocation.world)
            if (snapped != null && snapped.distanceSquared(nextLocation) < 4.0) {
                nextLocation = snapped
                nextX = snapped.x
                nextY = snapped.y
                nextZ = snapped.z
            }
        }

        nextLocation = RailPathUtil.project(nextLocation)
        nextX = nextLocation.x
        nextY = nextLocation.y
        nextZ = nextLocation.z

        return LeadState(
            nextX,
            nextY,
            nextZ,
            correctedVelocity.x,
            correctedVelocity.y,
            correctedVelocity.z,
            nextLocation,
        )
    }
}
